src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -1505,39 +1505,27 @@ public static bool CheckSchemeName(string? schemeName)
             (uint)(digit - 'a') <= 'f' - 'a' ? digit - 'a' + 10 :
             throw new ArgumentException(nameof(digit));
 
-        //
-        // GetHashCode
-        //
-        //  Overrides default function (in Object class)
-        //
-        //
         public override int GetHashCode()
         {
             if (IsNotAbsoluteUri)
             {
-                return CalculateCaseInsensitiveHashCode(OriginalString);
-            }
-
-            // Consider moving hash code storage from m_Info.MoreInfo to m_Info
-            UriInfo info = EnsureUriInfo();
-            if ((object?)info.MoreInfo == null)
-            {
-                info.MoreInfo = new MoreInfo();
+                MoreInfo moreInfo = (_info ??= new UriInfo()).MoreInfo ??= new MoreInfo();
```

Comment:
There is a dedicated field on the MoreInfo for caching this hash, it just wasn't being stored for relative Uris before.

Having a cached hash can also be used as another fast-path optimization in Equals.

> There's no need to complicate the code path by caching the result. string.GetHashCode() is already fairly efficient

Should we also not cache it for AbsoluteUris? If so, we can remove that field to save a pointer size on the allocation by trading GetHashCode/Equals perf.

Reply:
> There is a dedicated field on the MoreInfo for caching this hash

I'd guess that this caching mechanism wasn't intended to work around any perceived slowness in `string.GetHashCode` as much as it was to work around the possible allocation in the below code path.

https://github.com/dotnet/runtime/blob/106eebda0f91acb4354a3a83b5be5d70343f92f6/src/libraries/System.Private.Uri/src/System/Uri.cs#L4946-L4949

For absolute URIs there's still the possibility for allocation in the hash code computation routine, right?

Reply:
> For absolute URIs there's still the possibility for allocation in the hash code computation routine, right?

Yes, but only once. GetHashCode and Equals calls will cache the value in MoreInfo.RemoteUrl if/when they compute it.

Reply:
> I'd guess that this caching mechanism wasn't intended to work around any perceived slowness in string.GetHashCode as much as it was to work around the possible allocation in the below code path.

I don't know why it was not cached for the relative uri path, I assume it's because Info and MoreInfo are not used for anything by relative Uris otherwise.

Reply:
I removed the caching of the hash.